File: flink-test/src/main/java/com/atguigu/test/WordCount.java
package com.atguigu.test;

import com.sun.org.apache.xalan.internal.xsltc.compiler.util.Type;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.operators.AggregateOperator;
import org.apache.flink.api.java.operators.DataSource;
import org.apache.flink.api.java.operators.FlatMapOperator;
import org.apache.flink.api.java.tuple.Tuple;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.util.Collector;

import java.util.Collection;


/**
 * @ClassName WordCount
 * @Description TODO
 * @Author Xing
 * @Date 2021/4/8 16:45
 * @Version 1.0
 */
public class WordCount {
    public static void main(String[] args) throws Exception {
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
        DataSource<String> lineDS = env.readTextFile("");

        FlatMapOperator<String, Tuple2<String, Long>> wordAndOne = lineDS.flatMap(
                (String line, Collector<Tuple2<String, Long>> out) -> {
                    String[] strings = line.split(" ");
                    for (String word : strings) {
                        out.collect(Tuple2.of(word, 1L));
                    }
                }).returns(Types.TUPLE(Types.STRING, Types.LONG));

        AggregateOperator<Tuple2<String, Long>> sum = wordAndOne.groupBy(0).sum(1);
        sum.print();
    }
}

File: flink-scala/src/main/scala/com/atguigu/wc/WordCount.scala
package com.atguigu.wc

import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows

/**
  * @ClassName WordCount
  * @Description TODO
  * @Author Xing
  * @Date 2021/4/9 20:10
  * @Version 1.0
  */
object WordCount {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    val textDS: DataStream[String] = env.socketTextStream("hadoop102",9999)

    val result: DataStream[(String, Int)] = textDS.flatMap(_.split(" "))
      .map((_, 1))
      .keyBy(_._1)
      .window(TumblingProcessingTimeWindows.of(Time.seconds(5)))
      .sum(1)
    result.print()

    env.execute("Window Stream WordCount")

  }

}
